{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2bbc7088",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8244bc95",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Unitig:\n",
    "    def __init__(self, head, sequence, anomaly):\n",
    "        self.length = len(sequence)\n",
    "        self.head = head\n",
    "        self.sequence = sequence\n",
    "        self.anomaly = anomaly\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f\"{self.head}; len: {self.length} and A: {self.anomaly}\"\n",
    "    \n",
    "    @staticmethod\n",
    "    def extract(path, unitigs):\n",
    "        file = open(path, 'w')\n",
    "        for unitig in unitigs:\n",
    "            file.write(unitig.head + '\\n')\n",
    "            file.write(unitig.sequence + '\\n')\n",
    "        file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b7a8787",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_path = \"/home/Users/ns58/KOMB2/Data/PRJNA878603_SraRunTable.txt\"\n",
    "\n",
    "bucket = {}  # disease length : [SRA IDs]\n",
    "sample_to_condition = {} # SRA ID : disease length\n",
    "\n",
    "with open(info_path, \"r\") as file:\n",
    "    lines = file.readlines()\n",
    "    iterable = iter(lines)\n",
    "    next(iterable)  # skip first entry\n",
    "    for line in iterable:\n",
    "        data = line[:-1].split(',')  # 0: SRA ID, 24: gender, 29: disease_length\n",
    "        if len(data) < 30:\n",
    "            print(\"ERROR\")\n",
    "            continue\n",
    "        if data[29] not in bucket:\n",
    "            bucket[data[29]] = []\n",
    "        bucket[data[29]].append(data[0])\n",
    "        sample_to_condition[data[0]] = data[29]\n",
    "        \n",
    "samples = bucket['Control'] + bucket['Long'] + bucket['Short']\n",
    "sample_paths = [f\"/home/Users/ns58/KOMB2/PRJNA878603-Output/{sample}-k51-l150-c2\" for sample in samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6fb9d412",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting unitigs for BLAST. In order to run, remove this line\n"
     ]
    }
   ],
   "source": [
    "%%script echo 'Extracting unitigs for BLAST. In order to run, remove this line'\n",
    "\n",
    "top_unitigs_for_blast = []\n",
    "\n",
    "for sample, path in zip(samples, sample_paths):\n",
    "    df = pd.read_csv(f\"{path}/kcore.tsv\", sep='\\t')\n",
    "    anomaly_df = pd.read_csv(f\"{path}/CoreA_anomaly.txt\", sep='\\t', names=['#VID', 'Score'])\n",
    "    combined = pd.merge(df, anomaly_df, on='#VID')\n",
    "    \n",
    "    name_to_score = combined.set_index('Name')['Score'].to_dict()\n",
    "    \n",
    "    unitigs = []\n",
    "    \n",
    "    file = open(f\"{path}/combined.fasta\", 'r')\n",
    "    lines = file.readlines()\n",
    "    file.close()\n",
    "    \n",
    "    for idx in range(0, len(lines), 2):\n",
    "        first_part = lines[idx].split('|')[0]\n",
    "        name = int(first_part.split('_')[1])\n",
    "        \n",
    "        sequence = lines[idx + 1][:-1]\n",
    "        \n",
    "        unitigs.append(Unitig(f\">Unitig_{name}, from sample {sample}\", sequence, name_to_score[name]))\n",
    "        \n",
    "    unitigs.sort(key= lambda unitig:unitig.anomaly, reverse=True)\n",
    "    \n",
    "    top_unitigs_for_blast.extend(unitigs[:100])\n",
    "\n",
    "Unitig.extract(\"/home/Users/mt102/shared/KOMB-revision/BLAST-unitigs-100ps.fasta\", top_unitigs_for_blast)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59504138",
   "metadata": {},
   "source": [
    "### Create cvs (dataframe) containing info about each individual sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0f6f2a33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping\n"
     ]
    }
   ],
   "source": [
    "%%script echo 'Skipping'\n",
    "\n",
    "reads_cts = []\n",
    "# reads_2_cts = []\n",
    "unitigs_cts = []\n",
    "filtered_percents = []\n",
    "post_filter_cts = []\n",
    "conditions = []\n",
    "anomalous_percents = []\n",
    "\n",
    "reads_paths = [f\"/home/Users/ns58/KOMB2/Data/PRJNA878603-ME_CFS-NovaSeq/{sample}_1.fastq\" for sample in samples]\n",
    "# reads_2_paths = [f\"/home/Users/ns58/KOMB2/Data/PRJNA878603-ME_CFS-NovaSeq/{sample}_2.fastq\" for sample in samples]\n",
    "\n",
    "for sample, reads_path, path in zip(samples, reads_paths, sample_paths):\n",
    "    reads_ct = 0\n",
    "    # reads_2_ct = 0\n",
    "    with open(reads_path, 'r') as file:\n",
    "        for line in file.readlines():\n",
    "            if line[0] == '@':\n",
    "                reads_ct += 1\n",
    "    \n",
    "#     with open(reads_2_path, 'r') as file:\n",
    "#         for line in file.readlines():\n",
    "#             if line[0] == '@':\n",
    "#                 reads_2_ct += 1\n",
    "    \n",
    "    reads_cts.append(reads_ct)\n",
    "#    reads_2_cts.append(reads_2_ct)\n",
    "    \n",
    "    unitigs_ct = 0\n",
    "    \n",
    "    with open(f\"{path}/unitigs.fasta\", 'r') as file:\n",
    "        for line in file.readlines():\n",
    "            if line[0] == '>':\n",
    "                unitigs_ct += 1\n",
    "    \n",
    "    unitigs_cts.append(unitigs_ct)\n",
    "    \n",
    "    post_filter_ct = 0\n",
    "    \n",
    "    with open(f\"{path}/unitigs.l150.fasta\", 'r') as file:\n",
    "        for line in file.readlines():\n",
    "            if line[0] == '>':\n",
    "                post_filter_ct += 1\n",
    "    \n",
    "    post_filter_cts.append(post_filter_ct)\n",
    "    \n",
    "    filtered_percents.append(100*(1 - post_filter_ct / unitigs_ct))\n",
    "    \n",
    "    conditions.append(sample_to_condition[sample])\n",
    "    \n",
    "    anomaly_df = pd.read_csv(f\"{path}/CoreA_anomaly.txt\", sep='\\t', names=['#VID', 'Score'])\n",
    "    \n",
    "    mean_score = anomaly_df['Score'].mean()\n",
    "    std_score = anomaly_df['Score'].std()\n",
    "\n",
    "    # Compute the threshold value\n",
    "    threshold = mean_score + 3 * std_score\n",
    "\n",
    "    # Use the condition directly on the 'Score' column and count True values\n",
    "    num_entries_above_threshold = (anomaly_df['Score'] > threshold).sum()\n",
    "    \n",
    "    anomalous_percents.append(100.0 * num_entries_above_threshold / post_filter_ct)\n",
    "\n",
    "summary_df = pd.DataFrame({'Sample': samples, 'Diseased condition': conditions, \n",
    "                           'Reads count':reads_cts, 'Unitigs count': unitigs_cts,\n",
    "                           'Post-filtering unitigs count': post_filter_cts,\n",
    "                           'Filtered unitigs (%)':filtered_percents,\n",
    "                           'Anomalous out of filtered (%)': anomalous_percents})\n",
    "summary_df\n",
    "summary_df.to_csv(\"/home/Users/mt102/shared/KOMB-revision/CFS-data-summary.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "da5ef443",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Diseased condition</th>\n",
       "      <th>Reads count</th>\n",
       "      <th>Unitigs count</th>\n",
       "      <th>Post-filtering unitigs count</th>\n",
       "      <th>Filtered unitigs (%)</th>\n",
       "      <th>Anomalous out of filtered (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>SRR24907820</td>\n",
       "      <td>Control</td>\n",
       "      <td>4083896</td>\n",
       "      <td>1245343</td>\n",
       "      <td>333558</td>\n",
       "      <td>73.215572</td>\n",
       "      <td>0.503960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SRR24907827</td>\n",
       "      <td>Control</td>\n",
       "      <td>7995216</td>\n",
       "      <td>2417416</td>\n",
       "      <td>623255</td>\n",
       "      <td>74.218132</td>\n",
       "      <td>0.941188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>SRR24907851</td>\n",
       "      <td>Control</td>\n",
       "      <td>5644929</td>\n",
       "      <td>1236242</td>\n",
       "      <td>406294</td>\n",
       "      <td>67.134752</td>\n",
       "      <td>0.478964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>SRR24907866</td>\n",
       "      <td>Control</td>\n",
       "      <td>7136575</td>\n",
       "      <td>1335278</td>\n",
       "      <td>378976</td>\n",
       "      <td>71.618195</td>\n",
       "      <td>0.583942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>SRR24907869</td>\n",
       "      <td>Control</td>\n",
       "      <td>12846600</td>\n",
       "      <td>1568241</td>\n",
       "      <td>583632</td>\n",
       "      <td>62.784291</td>\n",
       "      <td>0.921814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>233</td>\n",
       "      <td>SRR24908020</td>\n",
       "      <td>Short</td>\n",
       "      <td>10397835</td>\n",
       "      <td>1364798</td>\n",
       "      <td>354630</td>\n",
       "      <td>74.015935</td>\n",
       "      <td>0.779686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>234</td>\n",
       "      <td>SRR24908028</td>\n",
       "      <td>Short</td>\n",
       "      <td>30568288</td>\n",
       "      <td>3714117</td>\n",
       "      <td>809333</td>\n",
       "      <td>78.209276</td>\n",
       "      <td>1.102883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>235</td>\n",
       "      <td>SRR24908035</td>\n",
       "      <td>Short</td>\n",
       "      <td>28107257</td>\n",
       "      <td>2955176</td>\n",
       "      <td>679237</td>\n",
       "      <td>77.015345</td>\n",
       "      <td>0.783968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>236</td>\n",
       "      <td>SRR24908039</td>\n",
       "      <td>Short</td>\n",
       "      <td>2593754</td>\n",
       "      <td>705590</td>\n",
       "      <td>207233</td>\n",
       "      <td>70.629828</td>\n",
       "      <td>0.505228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>237</td>\n",
       "      <td>SRR24908054</td>\n",
       "      <td>Short</td>\n",
       "      <td>5848398</td>\n",
       "      <td>1378346</td>\n",
       "      <td>446660</td>\n",
       "      <td>67.594494</td>\n",
       "      <td>0.930238</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>238 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0       Sample Diseased condition  Reads count  Unitigs count  \\\n",
       "0             0  SRR24907820            Control      4083896        1245343   \n",
       "1             1  SRR24907827            Control      7995216        2417416   \n",
       "2             2  SRR24907851            Control      5644929        1236242   \n",
       "3             3  SRR24907866            Control      7136575        1335278   \n",
       "4             4  SRR24907869            Control     12846600        1568241   \n",
       "..          ...          ...                ...          ...            ...   \n",
       "233         233  SRR24908020              Short     10397835        1364798   \n",
       "234         234  SRR24908028              Short     30568288        3714117   \n",
       "235         235  SRR24908035              Short     28107257        2955176   \n",
       "236         236  SRR24908039              Short      2593754         705590   \n",
       "237         237  SRR24908054              Short      5848398        1378346   \n",
       "\n",
       "     Post-filtering unitigs count  Filtered unitigs (%)  \\\n",
       "0                          333558             73.215572   \n",
       "1                          623255             74.218132   \n",
       "2                          406294             67.134752   \n",
       "3                          378976             71.618195   \n",
       "4                          583632             62.784291   \n",
       "..                            ...                   ...   \n",
       "233                        354630             74.015935   \n",
       "234                        809333             78.209276   \n",
       "235                        679237             77.015345   \n",
       "236                        207233             70.629828   \n",
       "237                        446660             67.594494   \n",
       "\n",
       "     Anomalous out of filtered (%)  \n",
       "0                         0.503960  \n",
       "1                         0.941188  \n",
       "2                         0.478964  \n",
       "3                         0.583942  \n",
       "4                         0.921814  \n",
       "..                             ...  \n",
       "233                       0.779686  \n",
       "234                       1.102883  \n",
       "235                       0.783968  \n",
       "236                       0.505228  \n",
       "237                       0.930238  \n",
       "\n",
       "[238 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df = pd.read_csv(\"/home/Users/mt102/shared/KOMB-revision/CFS-data-summary.tsv\", sep='\\t')\n",
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa7ff7dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7080a3df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c1f30f2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
